### ISSUE TITLE
Possible bug in `Linear` flops counter

### ISSUE THREAD

**Issue by forresti** [sovrasov/flops-counter.pytorch#28]
As I understand it, in PyTorch, `Linear` is the name for a fully-connected layer. Its forward pass is a matrix multiplication of an input tensor of shape [B, M, K] by a weight tensor of shape [B, K, N], with an output with shape of [B, M, K]. The number of MACs (not counting any bias terms or other things) is `B*M*N*K`. However, when I was using `flops-counter.pytorch` in my work, I noticed that `flops-counter.pytorch`'s MAC estimates for Linear layers seemed really low relative to my hand-calculations. I think I found the issue. The current code (below) only does `B*K*N`, and it does not account for the M dimension.
```
def linear_flops_counter_hook(module, input, output):
    input = input[0]
    batch_size = input.shape[0]
    module.__flops__ += int(batch_size * input.shape[1] * output.shape[1])
```

The following is my proposed fix. At least on my particular models, it produces results that match both my intuition and my calculations. If it would be of use, I would be happy to submit the following as a pull request.
```
def linear_flops_counter_hook(module, input, output):
    input = input[0]
    M = input.shape[-2]
    K = input.shape[-1]
    N = output.shape[-1]
    batch_size = input.shape[0]
    module.__flops__ += int(batch_size * M*N*K)
```

**Comment by sovrasov** [comment]
That's a bug indeed. Thanks for figuring it out!
I supposed that Linear takes a batch of N-dimensional vectors and then applies a linear transformation W with weights shape [M, N] to the each of these vectors. Then complexity is batch_size * M * N. When I build my models I always reshape the input to B*N before feeding it to a linear layer and in that case ptflops works correctly.

According to the documentation, pytorch allows inputs and outputs of the Linear layer to have any amount of dimensions and the weights tensor has shape [M, N]. pytorch applies linear transformation to the last dimension only. So we have input [B, D1, D2, ... D_s, N], output [B, D1, D2, ... D_s, M], weights [M, N]. Therefore complexity is B * D * N * M, where D=D1 * D2 * ... * D_s.

My implementation supposes s = 0, your fix supposes s = 1, but generally we need a common fix here to support any s >= 0.

PS. Your SqueezeNet is historical stuff, thanks for that work! It very impacted me when I started to study CNNs.

**Comment by sovrasov** [comment]
I've submitted a fix (#29). Please check if it works for you.

**Comment by forresti** [comment]
Great! I added #29 to my working-copy, and it works for my current pytorch models. 

P.S. Nice meeting you! Glad you enjoyed SqueezeNet.
